### ISSUE TITLE
Wrong timeout value for ExternalTaskSensor running in deferrable mode

### ISSUE THREAD

**Issue by kien-truong** [apache/airflow#43948]
### Apache Airflow version

2.10.3

### What happened?

The WorkflowTrigger used by ExternalTaskSensor should have a time limit set from `timeout` attribute instead of `execution_timeout`

https://github.com/apache/airflow/blob/bb234dc316f8421760255d85a37a10fd508de14f/airflow/sensors/external_task.py#L349

### How to reproduce

1. Start an ExternalTaskSensor in deferrable mode to monitor a non-running DAG, using default arguments.
2. The Sensor never timeout because the monitored DAGs do not update, and the timeout value is wrongly set from `execution_timeout` instead of `timeout`

### Operating System

Linux

### Deployment

Google Cloud Composer

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Assigned you

**Comment by karakanb** [comment]
I looked into this a bit but it seems like there's a fundamental issue here, I'll try to explain below.


The expected behavior would be to have a sensor that can run with retries, in case something fails during the sensor check, e.g. infra issues. The retries are not about the sensor not finding what it was supposed to, e.g. "the task is not there", but to recover from infra failures, e.g. the database being temporarily unavailable. This behavior works as expected with sensors in general.

However, when combining retries on sensors with timeouts, that's where things start getting interesting:
- When the user sets a timeout, the intention is "wait this long _from the beginning of the first try_", which is a very important factor that is also highlighted in the [Timeouts section of the docs](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/tasks.html#timeouts). This behavior seems to work correctly with `reschedule` mode thanks to the `task_reschedule` table that records the start timestamp for the first try.
- However, when deferrable mode is used, the timeouts do not work with retries since there's no way to retrieve the start time of the first attempt of a task instance.

It seems like the user would want the same behavior between deferred and non-deferred versions of the sensor for the timeouts with retries, but I couldn't find a way to solve it without adding a new table to airflow. is the original first start time information saved somewhere?

**Comment by kien-truong** [comment]
> When the user sets a timeout, the intention is "wait this long from the beginning of the first try

Yeah, even in `poke` mode sensor does not respect this behavior, but use the current run start_date as the starting point, which I think is a bug in itself.

**Comment by karakanb** [comment]
I think the document is correct with regards to what it says: it says "the `execution_timeout` and the `timeout` behave the same for `poke`", which means that they don't do what we'd expect them to do, and instead just set the timeout for the specific retry attempt instead.

I agree with you that they should behave the same way. I think it'd be a relatively simple fix, but there needs to be a state that stores the attempts.

**Comment by karakanb** [comment]
Actually, testing this out, it does not work indeed across attempts, instead the timeout is only enforced within the same attempt across different reschedules. I think @kien-truong is right.

**Comment by nathadfield** [comment]
There is currently a behavioural discrepancy between regular and deferred sensors that has been outstanding for many months which, it seems, will finally be [addressed](https://github.com/apache/airflow/pull/33718) in 2.11.x.

**Comment by nathadfield** [comment]
I think it is correct that sensors that have exhausted their timeout should not retry; instead the timeout should be longer to cover a wide enough window of waiting.

**Comment by seifrajhi** [comment]
Stills relevant, to remove stale label

**Comment by kien-truong** [comment]
It's partially overlap with [#33718](https://github.com/apache/airflow/pull/33718), which is making Deferrable sensor to not retry when `timeout` is reach. 

However, even with that fix in place, the sensor still needs to call defer with the correct `timeout` parameter, which is not the case with `ExternalTaskSensor`: it's passing `execution_timeout` instead of `timeout`

**Comment by nathadfield** [comment]
@dstandish Are you able to comment on this given you worked on #33718?
